templates/embeds/google-drive-file.php

```diff
@@ -9,10 +9,37 @@
 	// prevent direct access to this file.
 	exit;
 }
-?>
-<div style="border: 1px solid #000; text-align: center;">
-	<a href="<?php echo esc_url( $drive_file_url ); ?>" title="<?php esc_attr_e( 'Open the Shared Document', 'rt-google-embeds' ); ?>" target="_blank" rel="noopener noreferrer" style="color: #cd2653;">
-		<?php esc_html_e( 'Open Shared Document.', 'rt-google-embeds' ); ?>
-	</a>
-	<img src="<?php echo esc_url( $thumbnail_url ); ?>" alt="<?php esc_attr_e( 'Shared Document Preview', 'rt-google-embeds' ); ?>" style="border: 1px solid #eee; margin: 15px auto; display: block;" />
-</div>
+
+$matches = array();
+preg_match( '/[-\w]{25,}/', $drive_file_url, $matches );
+if ( ! empty( $matches[0] ) ) {
+	$file_id = $matches[0];
+}
+
+$upload_dir = wp_upload_dir();
+$basedir    = '';
+$baseurl    = '';
+
+if ( ! empty( $upload_dir['basedir'] ) && ! empty( $upload_dir['baseurl'] ) ) {
+	$basedir = $upload_dir['basedir'];
+	$baseurl = $upload_dir['baseurl'];
+}
+
+$cached_file = $basedir . "/cache/wp-google-drive/{$file_id}.png";
+$cached_url  = $baseurl . "/cache/wp-google-drive/{$file_id}.png";
```

Comment:
This is not supposed to be implemented here. This template is being used to print the image. You need to change the URL which is being passed in this template with locally cached image URL.

You need to implement this in `get_thumbnail_url` function. Request for the image online first, if it exists there take that image, else check if its cached locally and serve that.